hls4ml/model/optimizer/passes/infer_precision.py

```diff
@@ -0,0 +1,290 @@
+        # this in config_from_* functions
+
+        return []
+
+    def _infer_default_type(self, node, type_name):
+        model_config = node.model.config
+        default_precision = model_config.backend.convert_precision_string(model_config.model_precision['default'])
+        # No need to change the name of the NamedType since we use the default precision
+        node.types[type_name].precision = default_precision
+
+    def _infer_output_matching_precision(self, node, types_to_infer):
+        assert 'result_t' in types_to_infer and len(types_to_infer) == 1
+
+        in_var = node.get_input_variable()
+        out_var = node.get_output_variable()
+        in_out_type = in_var.type.precision
+        out_var.type.precision = in_out_type
+
+        return ['result_t']
+
+    def _infer_common_precision(self, node, types_to_infer, n_ops):
+        inferred_types = []
+
+        input_precision = node.get_input_variable().type.precision
+        input_width = input_precision.width
+        input_integers = input_precision.integer
+
+        if 'weight_t' in types_to_infer:
+            weight_quantizer = node.get_attr('weight_quantizer', None)
+            if weight_quantizer is not None:
+                weight_width = weight_quantizer.bits
+                weight_integers = weight_quantizer.hls_type.integer
+                node.types['weight_t'].name = node.name + '_weight_t'
+                node.types['weight_t'].precision = weight_quantizer.hls_type
+            else:
+                self._infer_default_type(node, 'weight_t')
+                weight_width = node.types['weight_t'].precision.width
+                weight_integers = node.types['weight_t'].precision.integer
+            node.weights['weight'].update_precision(node.types['weight_t'].precision)
+
+            inferred_types.append('weight_t')
+        else:
+            weight_width = node.types['weight_t'].precision.width
+            weight_integers = node.types['weight_t'].precision.integer
+
+        if 'bias_t' in types_to_infer:
+            bias_quantizer = node.get_attr('bias_quantizer', None)
+            if bias_quantizer is not None:
+                bias_width = bias_quantizer.bits
+                bias_integers = bias_quantizer.hls_type.integer
+                node.types['bias_t'].name = node.name + '_bias_t'
+                node.types['bias_t'].precision = bias_quantizer.hls_type
+            else:
+                self._infer_default_type(node, 'bias_t')
+                bias_width = node.types['bias_t'].precision.width
+                bias_integers = node.types['bias_t'].precision.integer
+            node.weights['bias'].update_precision(node.types['bias_t'].precision)
+
+            inferred_types.append('bias_t')
+        else:
+            bias_width = node.types['bias_t'].precision.width
+            bias_integers = node.types['bias_t'].precision.integer
+
+        new_type = FixedPrecisionType(
```

Comment:
Should we worry about unsigned inputs (or less likely, weights or biases)? The old version explicitly handled signedness.

Reply:
If I say we shouldn't, someone will come tomorrow with unsigned inputs, so perhaps we should consider doing it. I'll update based on your code.

Reply:
Just for concreteness, here is the precision inference we used when we tried to match QKeras at the CMS Trigger workshop:
```python
def _propagate_type_dense(input_precision, weight_precision, bias_precision, num_acc):
    '''
    Propagate the precion type across a multiply. Rounding modes are propagated from input_precision
    '''

    # check to make sure none are None
    bitwidth = weight_precision.width + input_precision.width + math.ceil(np.log2(num_acc))
    integer = weight_precision.integer + input_precision.integer + math.ceil(np.log2(num_acc))
    signed = weight_precision.signed or input_precision.signed

    # Because calculating precision, no need to round or satration
    rounding_mode = None
    saturation_mode = None

    frac = bitwidth - integer

    # correct for bias
    if bias_precision:
        integer = max(integer + (bias_precision.signed and not signed),
                      bias_precision.integer + (signed and not bias_precision.signed)) + 1
        bitwidth = integer + max(frac, bias_precision.width - bias_precision.integer)
        signed = signed or bias_precision.signed

    return FixedPrecisionType(bitwidth, integer, signed, rounding_mode, saturation_mode)
```